File: snippets/fsharp/VS_Snippets_CLR_System/system.DBNull.Class/fs/DBNullExamples.fs
open System
open System.Data
open System.Data.OleDb

type DBNullExample() =
    // <Snippet1>
    member this.OutputLabels(dt: DataTable) =
        let mutable label = ""

        // Iterate rows of table
        for row in dt.Rows do
            let mutable label = String.Empty
            label <- label + this.AddFieldValue(label, row, "Title")
            label <- label + this.AddFieldValue(label, row, "FirstName")
            label <- label + this.AddFieldValue(label, row, "MiddleInitial")
            label <- label + this.AddFieldValue(label, row, "LastName")
            label <- label + this.AddFieldValue(label, row, "Suffix")
            label <- label + "\n"
            label <- label + this.AddFieldValue(label, row, "Address1")
            label <- label + this.AddFieldValue(label, row, "AptNo")
            label <- label + "\n"
            let labelLen = label.Length
            label <- label + this.AddFieldValue(label, row, "Address2")
            let labelLen =
                if label.Length <> labelLen then
                    label + "\n"
                else label
            label <- label + this.AddFieldValue(label, row, "City")
            label <- label + this.AddFieldValue(label, row, "State")
            label <- label + this.AddFieldValue(label, row, "Zip")
            printfn $"{label}"
            printfn ""

    member _.AddFieldValue(label: string, row: DataRow, fieldName: string) =
        if DBNull.Value.Equals row[fieldName] |> not then
            (string row[fieldName]) + " "
        else
            String.Empty
    // </Snippet1>

let ex = DBNullExample()
let conn = new OleDbConnection()
let cmd = new OleDbCommand()
let adapter = new OleDbDataAdapter()
let ds = new DataSet()
let dbFilename = @"c:\Data\contacts.mdb"

// Open database connection
conn.ConnectionString <- "Provider=Microsoft.Jet.OLEDB.4.0Data Source=" + dbFilename + ""
conn.Open()
// Define command : retrieve all records in contact table
cmd.CommandText <- "SELECT * FROM Contact"
cmd.Connection <- conn
adapter.SelectCommand <- cmd
// Fill dataset
ds.Clear()
adapter.Fill(ds, "Contact")
|> ignore
// Close connection
conn.Close()
// Output labels to console	
ex.OutputLabels ds.Tables["Contact"]

File: snippets/csharp/System/DBNull/Overview/DBNullExamples.cs
using System;
using System.Data;
using System.Data.OleDb;
using System.Runtime.Versioning;

public class DBNullExample
{
    [SupportedOSPlatform("windows")]
    public static void Main()
    {
        DBNullExample ex = new();
        OleDbConnection conn = new();
        OleDbCommand cmd = new();
        OleDbDataAdapter adapter = new();
        DataSet ds = new();
        string dbFilename = @"c:\Data\contacts.mdb";

        // Open database connection
        conn.ConnectionString = $"Provider=Microsoft.Jet.OLEDB.4.0;Data Source={dbFilename};";
        conn.Open();
        // Define command : retrieve all records in contact table
        cmd.CommandText = "SELECT * FROM Contact";
        cmd.Connection = conn;
        adapter.SelectCommand = cmd;
        // Fill dataset
        ds.Clear();
        adapter.Fill(ds, "Contact");
        // Close connection
        conn.Close();
        // Output labels to console
        ex.OutputLabels(ds.Tables["Contact"]);
    }

    // <Snippet1>
    private void OutputLabels(DataTable dt)
    {
        string label;

        // Iterate rows of table
        foreach (DataRow row in dt.Rows)
        {
            int labelLen;
            label = string.Empty;
            label += AddFieldValue(row, "Title");
            label += AddFieldValue(row, "FirstName");
            label += AddFieldValue(row, "MiddleInitial");
            label += AddFieldValue(row, "LastName");
            label += AddFieldValue(row, "Suffix");
            label += "\n";
            label += AddFieldValue(row, "Address1");
            label += AddFieldValue(row, "AptNo");
            label += "\n";
            labelLen = label.Length;
            label += AddFieldValue(row, "Address2");
            if (label.Length != labelLen)
                label += "\n";
            label += AddFieldValue(row, "City");
            label += AddFieldValue(row, "State");
            label += AddFieldValue(row, "Zip");
            Console.WriteLine(label);
            Console.WriteLine();
        }
    }

    private string AddFieldValue(DataRow row,
                                 string fieldName)
    {
        if (!DBNull.Value.Equals(row[fieldName]))
            return (string)row[fieldName] + " ";
        else
            return string.Empty;
    }
    // </Snippet1>
}
